nca/Parsers/IstioGatewayPolicyGenerator.py

```diff
@@ -0,0 +1,298 @@
+
+    def __init__(self, gtw_parser, vs_parser):
+        """
+        :param IstioGatewayParser gtw_parser: the gateway parser containing the parsed gateways.
+        :param IstioVirtualServiceParser vs_parser: the virtual service parser containing the parsed virtual services.
+        """
+        self.gtw_parser = gtw_parser
+        self.vs_parser = vs_parser
+
+    def create_istio_gateway_policies(self):
+        """
+        Create GatewayPolicies according to the parsed Istio Gateways and VirtualServices
+        :return list[GatewayPolicy]: the resulting policies
+        """
+        if not self.gtw_parser or not self.gtw_parser.gateways:
+            self.vs_parser.warning('no valid Gateways found. Ignoring istio ingress/egress traffic')
+            return []
+
+        if not self.vs_parser or not self.vs_parser.virtual_services:
+            self.gtw_parser.warning('no valid VirtualServices found. Ignoring istio ingress/egress traffic')
+            return []
+
+        result = []
+        used_gateways = set()
+        for vs in self.vs_parser.virtual_services.values():
+            global_vs_gtw_to_hosts = self.pick_vs_gateways_by_hosts(vs, vs.gateway_names)
+            used_gateways.update(set(global_vs_gtw_to_hosts.keys()))
+            only_local_traffic = False
+            vs_policies = self.create_route_policies(vs, vs.http_routes, global_vs_gtw_to_hosts, used_gateways,
+                                                     only_local_traffic)
+            vs_policies.extend(self.create_route_policies(vs, vs.tls_routes, global_vs_gtw_to_hosts, used_gateways,
+                                                          only_local_traffic))
+            vs_policies.extend(self.create_route_policies(vs, vs.tcp_routes, global_vs_gtw_to_hosts, used_gateways,
+                                                          only_local_traffic))
+            if vs_policies:
+                result.extend(vs_policies)
+            else:
+                if only_local_traffic:
+                    self.vs_parser.warning(f'The virtual service {vs.full_name()} defines '
+                                           f'only local (mesh-to-mesh) traffic that is ignored', vs)
+                else:
+                    self.vs_parser.warning(f'The virtual service {vs.full_name()} does not affect traffic '
+                                           f'and is ignored', vs)
+
+        unused_gateways = self.gtw_parser.get_all_gateways() - used_gateways
+        if unused_gateways:
+            self.vs_parser.warning(f'the following gateways have no virtual services attached: '
+                                   f'{",".join([gtw.full_name() for gtw in unused_gateways])}')
+        if not result:
+            self.vs_parser.warning('no valid VirtualServices found. Ignoring istio ingress traffic')
+        else:
+            result[0].findings = self.gtw_parser.warning_msgs + self.vs_parser.warning_msgs
```

Comment:
Unlike network policies, where a different parser is created to each policy, in case of gateway policies one parser is created for all gateways and one parser is created for all virtual services. Furthermore, there is no one-to-one correspondence between `GatewayPolicy` and `Gateway/VirtualService`. Thus, all warning messages are accumulated in the above two parsers (some warnings are accumulated during the parse phase, and others during gateway policy creation phase). It may be a future improvement, to differentiate warning messages according to a resource (`Gateway/VirtualService`) from which they originated.